src/librustdoc/html/static/js/search.js

```diff
@@ -133,6 +124,319 @@ window.initSearch = function(rawSearchIndex) {
+    function getStringElem(query, parserState, isInGenerics) {
+        if (isInGenerics) {
+            throw new Error("`\"` cannot be used in generics");
+        } else if (query.literalSearch) {
+            throw new Error("Cannot have more than one literal search element");
+        } else if (parserState.totalElems !== 0) {
+            throw new Error("Cannot use literal search when there is more than one element");
+        }
+        parserState.pos += 1;
+        while (parserState.pos < parserState.length &&
+            parserState.userQuery[parserState.pos] !== "\"")
+        {
+            parserState.pos += 1;
+        }
+        if (parserState.pos >= parserState.length) {
+            throw new Error("Unclosed `\"`");
+        }
+        // To skip the quote at the end.
+        parserState.pos += 1;
+        query.literalSearch = true;
+    }
+
+    /**
+     * Returns `true` if the current parser position is starting with "::".
+     *
+     * @param {ParserState} parserState
+     * @return {boolean}
+     */
+    function isPathStart(parserState) {
+        return parserState.userQuery.slice(parserState.pos, parserState.pos + 2) == '::';
+    }
+
+    /**
+     * Returns `true` if the current parser position is starting with "->".
+     *
+     * @param {ParserState} parserState
+     * @return {boolean}
+     */
+    function isReturnArrow(parserState) {
+        return parserState.userQuery.slice(parserState.pos, parserState.pos + 2) == '->';
+    }
+
+    /**
+     * @param {ParsedQuery} query
+     * @param {ParserState} parserState
+     * @param {Array<QueryElement>} elems    - This is where the new {QueryElement} will be added.
+     * @param {string} name                  - Name of the query element.
+     * @param {Array<QueryElement>} generics - List of generics of this query element.
+     */
+    function createQueryElement(query, parserState, elems, name, generics) {
+        if (name === '*' || (name.length === 0 && generics.length === 0)) {
+            return;
+        }
+        if (query.literalSearch && parserState.totalElems > 0) {
+            throw new Error("You cannot have more than one element if you use quotes");
+        }
+        var pathSegments = name.split("::");
+        if (pathSegments.length > 1) {
+            for (var i = 0, len = pathSegments.length; i < len; ++i) {
+                var pathSegment = pathSegments[i];
+
+                if (pathSegment.length === 0) {
+                    if (i === 0) {
+                        throw new Error("Paths cannot start with `::`");
+                    } else if (i + 1 === len) {
+                        throw new Error("Paths cannot end with `::`");
+                    }
+                    throw new Error("Unexpected `::::`");
+                }
+            }
+        }
+        // In case we only have something like `<p>`, there is no name but it remains valid.
+        if (pathSegments.length === 0) {
+            pathSegments = [""];
```

Comment:
```suggestion
            throw new Error("Search queries must contain a path");
```

Reply:
They seem to be supported currently though. Wouldn't that be a change in behaviour?

Reply:
When I type `<Clone>` [into the current stable docs](https://doc.rust-lang.org/stable/std/?search=%3CClone%3E), it doesn't seem to be checking for actual generics. The results look the same to me as they do when I just type `Clone`.

That seems more like a bug than an actual, intended behavior.

Reply:
Ok, I'll prevent this for now then!

Reply:
The more I think about it, the more I'm unsure about this change. cc @camelid @jsha 

Just in case, here's the diff:

```diff
--- a/src/librustdoc/html/static/js/search.js
+++ b/src/librustdoc/html/static/js/search.js
@@ -238,9 +238,9 @@ window.initSearch = function(rawSearchIndex) {
                 }
             }
         }
-        // In case we only have something like `<p>`, there is no name but it remains valid.
-        if (pathSegments.length === 0) {
-            pathSegments = [""];
+        // In case we only have something like `<p>`, there is no name.
+        if (pathSegments.length === 0 || (pathSegments.length === 1 && pathSegments[0] === "")) {
+            throw new Error("Found generics without a path");
```

Reply:
The [formal grammar](https://github.com/rust-lang/rust/pull/90630/files#diff-766c50b21af39c8c0abd3fdba5f3c49a73c38b7b66ded4c7939e83b7244dfd3dR482) does say that a type-search is defined in terms of path, and path is defined as non-empty. If we're going to allow empty paths, then that grammar should also be changed. to reflect this.

Reply:
I think we should require path to be non-empty for generics.

Reply:
Then I'll remove this possibility.